### DISCUSSION TITLE
Formily.Schema.Types.ISchema 类型中的索引签名是否可以去掉？

### DISCUSSION THREAD

**Discussion by liuweiGL** [alibaba/formily#1657] [Show and tell]
## 问题
在 Formily.Schema.Types.ISchema 中存在索引签名：
```ts
// ...
[key: string]: any;
```
使用 `Omit` 或则 `Pick` 扩展类型的时候丢失 ISchema  字段信息，导致 vscode 的 autocomplete 功能无法使用：

```
type Test = Omit<Formily.Schema.Types.ISchema, 'items'> & {
    items?: Test
    // ...
}
```

## 原因
`[key: string]: any` 返回的 `keyof` 为 `string`，把原有的字段覆盖了。

https://github.com/microsoft/TypeScript/issues/31153

**Comment by janryWang** [comment]
不能去掉，因为要支持挂自定义属性

**Reply by liuweiGL** [reply]
挂载自定义属性让用户在 Formily.Schema.Types.ISchema 类型上扩展呢？这样的话也更 `ts` 吧？

**Reply by janryWang** [reply]
不好扩展的，比如Markup Schema强依赖ISchema，用户是没法扩展的，只能创建一个联合新类型

**Reply by MisicDemone** [reply]
这个也是导致vue版本 markupField tsx没有代码提示的原因

**Reply by janryWang** [reply]
> 这个也是导致vue版本 markupField tsx没有代码提示的原因

卧槽，React也有这个问题，看来真得干掉了

**Comment by janryWang** [comment]
干掉了 https://github.com/alibaba/formily/commit/76a302cc5fcccb3cddf98616d047ceaf74d5a51e

**Comment by liuweiGL** [comment]
等 typescript@4.4 可以使用 template literal 做 index signature 时加回去：

```ts

[key: `x-${string}`]: any
```

https://github.com/microsoft/TypeScript/pull/44512
